thrift/src/main/java/com/linecorp/armeria/internal/client/thrift/THttpClientDelegate.java

```diff
@@ -139,21 +141,24 @@ public RpcResponse execute(ClientRequestContext ctx, RpcRequest call) throws Exc
                                   .authority(endpoint != null ? endpoint.authority() : "UNKNOWN")
                                   .contentType(mediaType)
                                   .build(),
-                    new ByteBufHttpData(buf, true));
+                    PooledHttpData.wrap(buf).withEndOfStream());
 
             ctx.updateRequest(httpReq);
             ctx.logBuilder().deferResponseContent();
 
-            final CompletableFuture<AggregatedHttpResponse> future =
-                    delegate().execute(ctx, httpReq).aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc());
+            assert delegate() instanceof PooledHttpClient;
+            final PooledHttpClient client = delegate();
+            final CompletableFuture<PooledAggregatedHttpResponse> future =
+                    client.execute(ctx, PooledHttpRequest.of(httpReq))
+                          .aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc());
 
             future.handle((res, cause) -> {
                 if (cause != null) {
                     handlePreDecodeException(ctx, reply, func, Exceptions.peel(cause));
                     return null;
                 }
 
-                try {
+                try (SafeCloseable unused = res) {
```

Comment:
This looks funny, just sayin' 😆

Reply:
Does give some credit to go's `defer` statement pattern :)